rfcs/0000-central-station.md

```diff
@@ -0,0 +1,507 @@
+new developers. Using a DB as the primary datastore should be easier for new developers, since
+most are accustomed to working with DBs.
+
+## What other designs have been considered and what is the rationale for not choosing them?
+
+The biggest alternative is to build on the status quo. Currently, all the modes work tolerably, if
+separately. Service Changes can be accommodated by Skate, Glides, and Keolis separately. We can
+lean on Swiftly for vehicle tagging.
+
+One downside of this is it locks us much more tightly into Swiftly. If we rely on Swiftly for
+tagging, and have several apps all using Swiftly's API, then we are tightly coupled to them,
+making it hard to move off them in the future (if our TransitClock experiments prove fruitful), or
+giving them leverage to raise prices quite a bit (like Google did with Maps). If Swiftly does a
+poor job of tagging vehicles, we have no recourse.
+
+If Swiftly moves quickly on ingesting ServiceChanges data rather than its API, then the coupling
+is a bit less here, but if they don't, and only enable it later on, we may have a variety of apps
+that will all need to switch from the API to ServiceChanges approach.
+
+Developers will continue to have to maintain working knowledge of, and actively build on and
+debug, a variety of similar yet different applications.
+
+Another alternative is a scaled-down CentralStation. It could be the centralized interface for
+Glides, Skate, and Keolis to post Service Changes, but not maintain the state of vehicles itself.
+This helps somewhat with the Swiftly coupling, but without having the vehicle and trip state, it
+can't validate the requests, or perhaps easily know when one has expired. It also doesn't support
+vehicle tagging, which is a major part of what these tools will do.
+
+## What is the impact of not doing this?
+
+The impact of not making a change here is that things will continue on under the status quo.
+Glides, Skate, and Keolis will soon be faced with making their changes, and will do so
+independently.
+
+Developers may feel burnt out working on lots of different apps.
+
+CTD and the MBTA may not have the best handle on our realtime data.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+WMATA is an example of a transit agency that has a centralized web-app powering a lot of its real
+time information. In a demo given to Jessie, Eddie, and me about single tracking, they pulled up a
+web page which showed the realtime position of various trains, as well as the state of switches.
+In the demo they used the web interface to manually indicate that a switch had changed, indicating
+single tracking, and showed how that affected predictions and signs downstream.
+
+Regarding the use of a DB here, in my past experience I've really appreciated postgres, and the
+data we work with is relational and fits naturally. However, in our experience at CTD with
+prediction analyzer, we can run into performance issues if we're not careful.
```

Comment:
Could performance issues be mitigated by running separate instances of CentralStation per mode? It seems like the benefits of consolidating all modes together are mostly at the development level, and keeping modes separate at the operational level would mean performance bottlenecks are isolated to a single mode.

Reply:
Service changes often cut across modes (e.g. a replacement of part of a subway line with shuttle buses). If there were separate data stores for each mode, they would need a way to communicate and stay in sync with each other, somewhat undoing the benefits of everything living under one roof.

Reply:
Ah, yeah, if changes frequently don't apply to modes independently then it'd probably be unhelpful to silo data by mode for performance reasons.